py/server/deephaven/barrage.py

```diff
@@ -0,0 +1,176 @@
+_remote_session_lock = threading.Lock()
+
+
+def barrage_session(host: str,
+                    port: int = 10000,
+                    auth_type: str = "Anonymous",
+                    auth_token: str = "",
+                    use_tls: bool = False,
+                    tls_root_certs: bytes = None,
+                    ) -> BarrageSession:
+    """Returns a Deephaven gRPC session to a remote server if a cached session is available; otherwise, creates a new
+    session.
+
+    Note: client authentication is not supported yet.
+
+    Args:
+        host (str): the host name or IP address of the Deephaven server.
+        port (int): the port number that the remote Deephaven server is listening on, default is 10000.
+        auth_type (str): the authentication type string, can be "Anonymous', 'Basic", or any custom-built
+            authenticator in the server, such as "io.deephaven.authentication.psk.PskAuthenticationHandler",
+            default is 'Anonymous'.
+        auth_token (str): the authentication token string. When auth_type is 'Basic', it must be
+            "user:password"; when auth_type is "Anonymous', it will be ignored; when auth_type is a custom-built
+            authenticator, it must conform to the specific requirement of the authenticator
+        use_tls (bool): if True, use a TLS connection.  Defaults to False
+        tls_root_certs (bytes): PEM encoded root certificates to use for TLS connection, or None to use system defaults.
+             If not None implies use a TLS connection and the use_tls argument should have been passed
+             as True. Defaults to None
+
+    Returns:
+        a Deephaven Barrage session
+
+    Raises:
+        DHError
+    """
+    try:
+        if tls_root_certs and not use_tls:
+            raise DHError(message="tls_root_certs is provided but use_tls is False")
+
+        target_uri = f"{host}:{port}"
+        if use_tls:
+            target_uri = f"dh://{target_uri}"
+        else:
+            target_uri = f"dh+plain://{target_uri}"
+
+        j_client_config = _build_client_config(target_uri, tls_root_certs)
+        auth = f"{auth_type} {auth_token}"
+
+        try:
+            return _get_barrage_session_uri(j_client_config, auth)
+        except:
+            # fall back to the direct way when we don't have a fully initialized server, used for testing
+            return _get_barrage_session_direct(j_client_config, auth)
+    except Exception as e:
+        raise DHError(e, "failed to get a barrage session to the target remote Deephaven server.") from e
+
+
+def _get_barrage_session_uri(client_config, auth) -> BarrageSession:
+    j_barrage_session_factory_client = uri.resolve(
+        "dh:///app/io.deephaven.server.barrage.BarrageSessionFactoryClient/field/instance")
+    j_barrage_session_factory = j_barrage_session_factory_client.factory(client_config, auth)
+    return BarrageSession(j_barrage_session_factory.newBarrageSession())
```

Comment:
So, we may actually need some additional logic here, otherwise we'll be leaking ManagedChannels; we probably need to pass through `j_barrage_session_factory.managedChannel()` and shut it down as appropriate.